modules/calib3d/src/stereobm.cpp

```diff
@@ -846,13 +849,30 @@ findStereoCorrespondenceBM( const Mat& left, const Mat& right,
                 sad[ndisp] = sad[ndisp-2];
                 int p = sad[mind+1], n = sad[mind-1];
                 d = p + n - 2*sad[mind] + std::abs(p - n);
-                dptr[y*dstep] = (short)(((ndisp - mind - 1 + mindisp)*256 + (d != 0 ? (p-n)*256/d : 0) + 15) >> 4);
+                dptr[y*dstep] = (mType)(((ndisp - mind - 1 + mindisp)*256 + (d != 0 ? (p-n)*256/d : 0) + 15)
+                                 >> (DISPARITY_SHIFT_32S - DISPARITY_SHIFT));
                 costptr[y*coststep] = sad[mind];
             }
         }
     }
 }
 
+static void
+findStereoCorrespondenceBM( const Mat& left, const Mat& right,
+                            Mat& disp, Mat& cost, const StereoBMParams& state,
+                            uchar* buf, int _dy0, int _dy1 )
+{
+    if(disp.type() == CV_16S){
+        DISPARITY_SHIFT = DISPARITY_SHIFT_16S;
+        findStereoCorrespondenceBM_<short>(left, right, disp, cost, state,
+                                           buf, _dy0, _dy1 );
+    } else {
+        DISPARITY_SHIFT = DISPARITY_SHIFT_32S;
```

Comment:
currently, the function supports CV_16S and CV_32F datatypes for the disparity; It looks like you are using CV_32S here instead of CV_32F. It will likely break things.